File: solutions/cpp/remove-element.cc
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <algorithm>
#include <cassert>
#include <iostream>
#include <random>
#include <string>
#include <vector>

using std::cout;
using std::default_random_engine;
using std::endl;
using std::random_device;
using std::stoul;
using std::uniform_int_distribution;
using std::vector;

// @include
// Returns the number of valid entries after deletion.
size_t DeleteKey(int key, vector<int>* A_ptr) {
  auto& A = *A_ptr;
  size_t write_idx = 0;
  for (size_t i = 0; i < A.size(); ++i) {
    if (A[i] != key) {
      A[write_idx++] = A[i];
    }
  }
  return write_idx;
}
// @exclude

void CheckAns(const vector<int>& A, size_t n, int key) {
  for (size_t i = 0; i < n; ++i) {
    assert(A[i] != key);
  }
}

int main(int argc, char** argv) {
  default_random_engine gen((random_device())());
  size_t n;
  if (argc == 2) {
    n = stoul(argv[1]);
  } else {
    uniform_int_distribution<size_t> dis_n(0, 10000);
    n = dis_n(gen);
  }
  for (int times = 0; times < 1000; ++times) {
    uniform_int_distribution<int> A_dis(-1000, 1000);
    vector<int> A;
    generate_n(back_inserter(A), n, [&] { return A_dis(gen); });
    auto copy_A(A);
    int target = A_dis(gen);
    auto size = DeleteKey(target, &A);
    cout << "size = " << size << " key = " << target << endl;
    CheckAns(A, size, target);
    auto it = remove_if(copy_A.begin(), copy_A.end(),
                        [&target](int a) { return a == target; });
    cout << distance(copy_A.begin(), it) << endl;
    assert(size == distance(copy_A.begin(), it));
  }
  return 0;
}

File: solutions/python/remove_element.py
import sys
import random


# @include
# Returns the number of valid entries after deletion.
def delete_key(k, A):
    write_idx = 0
    for a in A:
        if a != k:
            A[write_idx] = a
            write_idx += 1
    return write_idx
# @exclude


def delete_key_pythonic(k, A):
    A[:] = [a for a in A if a != k]
    return len(A)


def check_ans(A, n, k):
    for i in range(n):
        assert A[i] != k


def main():
    n = int(sys.argv[1]) if len(sys.argv) == 2 else random.randint(0, 10000)
    for _ in range(1000):
        A = [random.randint(-1000, 1000) for _ in range(n)]
        copy_A = A.copy()
        target = random.randint(-1000, 1000)
        size = delete_key(target, A)
        print('size =', size, 'key =', target)
        check_ans(A, size, target)
        delete_key_pythonic(target, copy_A)
        print(len(copy_A))
        assert size == len(copy_A)


if __name__ == '__main__':
    main()
